### PR TITLE
JAMES-3202 retrieve indexed flags

### PR DESCRIPTION
**Pull request by Arsnael**
First part result of splitting #3429

### CODE REVIEW

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java [hunk 1]
```diff
@@ -213,4 +218,40 @@ private void logIfNoMessageId(SearchResult searchResult) {
...
+            .filter(GetResponse::isExists)
+            .map(GetResponse::getSourceAsMap)
+            .map(this::extractFlags)
+            .switchIfEmpty(Mono.error(new IndexNotFoundException(
+                String.format("Index for message %s in mailbox %s not found", uid.toString(), mailbox.getMailboxId().serialize()))));
+    }
+
+    private Flags extractFlags(Map<String, Object> source) {
+        FlagsBuilder flagsBuilder = FlagsBuilder.builder()
+            .isAnswered(extractFlag(source, JsonMessageConstants.IS_ANSWERED))
```
Comment (nvduc91) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java:236]: could we import static?

File: mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java [hunk 2]
```diff
@@ -1297,4 +1294,51 @@ public void delete(MailboxId mailboxId, MessageRange range) throws IOException {
...
+                    .forEach(flag -> fromString(flag).ifPresentOrElse(retrievedFlags::add, () -> retrievedFlags.add(flag)));
+            }
+            return retrievedFlags;
+        }
+    }
+
+    /**
+     * Convert the given {@link Flag} to a String
+     */
+    private Optional<Flag> fromString(String flag) {
```
Comment (nvduc91) [inline mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java:1327]: why not make switch-case for this?

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java [hunk 3]
```diff
@@ -217,6 +221,37 @@ private void logIfNoMessageId(SearchResult searchResult) {
...
     @Override
     public Mono<Flags> retrieveIndexedFlags(Mailbox mailbox, MessageUid uid) {
-        return Mono.empty();
+        RoutingKey routingKey = routingKeyFactory.from(mailbox.getMailboxId());
+
+        return elasticSearchIndexer.get(indexIdFor(mailbox, uid), routingKey)
+            .filter(GetResponse::isExists)
+            .map(GetResponse::getSourceAsMap)
+            .map(this::extractFlags)
+            .switchIfEmpty(Mono.error(new IndexNotFoundException(
```
Comment (remk) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java:236]: Defer error Mono

File: mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ReIndexerImplTest.java [hunk 4]
```diff
@@ -98,7 +98,7 @@ void reIndexAllShouldCallMessageSearchIndex() throws Exception {
     }
 
     @Test
-    void reIndexMailboxPathShouldCallMessageSearchIndex() throws Exception {
+    void reIndexAllShouldCallMessageSearchIndex() throws Exception {
```
Comment (chibenwa) [inline mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ReIndexerImplTest.java:101]: Unrelated changes?

File: mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ReIndexerImplTest.java [hunk 5]
```diff
@@ -71,7 +71,7 @@ void setUp() {
     }
 
     @Test
-    void reIndexAllShouldCallMessageSearchIndex() throws Exception {
+    void reIndexMailboxPathShouldCallMessageSearchIndex() throws Exception {
```
Comment (chibenwa) [inline mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ReIndexerImplTest.java:74]: Unrelated changes?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/LazyMessageSearchIndex.java [hunk 6]
```diff
@@ -1,150 +1,157 @@
...
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(LazyMessageSearchIndex.class);
+    private static final Group GROUP = new LazyMessageSearchIndexGroup();
+
+    private final ListeningMessageSearchIndex index;
+    private final ConcurrentHashMap<MailboxId, Object> indexed = new ConcurrentHashMap<>();
+    private final MailboxSessionMapperFactory factory;
+    
+
```
Comment (LanKhuat) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/LazyMessageSearchIndex.java:76]: delete extra linebreak

File: mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ReIndexerImplTest.java (hunk 4 above)
Comment (Arsnael) [reply mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ReIndexerImplTest.java:101]: Test renaming... I saw some tests with wrong names along the way and included a quick fix while I was at it. Is this a problem?

File: mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ReIndexerImplTest.java (hunk 5 above)
Comment (Arsnael) [reply mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ReIndexerImplTest.java:74]: idem

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java [hunk 7]
```diff
@@ -442,4 +446,102 @@ void deleteAllShouldNotThrowWhenEmptyIndex() {
             .doesNotThrowAnyException();
     }
 
+    @Test
+    void retrieveIndexedFlagsShouldRetrieveSystemFlags() {
```
[the remark this answers is not in the document]
Comment (Arsnael) [reply mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java:450]: I understand you want a contract test (will work on it), but we are already using JUnit 5? Or did I misunderstood the comment?

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java [hunk 8]
```diff
@@ -233,8 +233,8 @@ private void logIfNoMessageId(SearchResult searchResult) {
             .filter(GetResponse::isExists)
             .map(GetResponse::getSourceAsMap)
             .map(this::extractFlags)
-            .switchIfEmpty(Mono.error(new IndexNotFoundException(
-                String.format("Index for message %s in mailbox %s not found", uid.toString(), mailbox.getMailboxId().serialize()))));
+            .switchIfEmpty(Mono.defer(() -> Mono.error(new IndexNotFoundException(
```
Comment (chibenwa) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java:236]: ` .switchIfEmpty(Mono.error(() -> new IndexNotFoundException(...`

**Comment by chibenwa** [comment]
Merged
